components/brave_federated_learning/brave_operational_profiling.cc

```diff
@@ -0,0 +1,205 @@
+  DCHECK(!simulate_local_training_step_timer_);
+  DCHECK(!collection_slot_periodic_timer_);
+
+  LoadParams();
+
+  simulate_local_training_step_timer_ =
+      std::make_unique<base::RetainingOneShotTimer>();
+  simulate_local_training_step_timer_->Start(
+      FROM_HERE,
+      base::TimeDelta::FromSeconds(
+          simulated_local_training_step_duration_in_minutes_ * 60),
+      this, &BraveOperationalProfiling::OnSimulateLocalTrainingStepTimerFired);
+
+  collection_slot_periodic_timer_ = std::make_unique<base::RepeatingTimer>();
+  collection_slot_periodic_timer_->Start(
+      FROM_HERE,
+      base::TimeDelta::FromSeconds(collection_slot_size_in_minutes_ * 60 / 2),
+      this, &BraveOperationalProfiling::OnCollectionSlotStartTimerFired);
+}
+
+void BraveOperationalProfiling::LoadParams() {
+  collection_slot_size_in_minutes_ =
+      operational_profiling::features::GetCollectionSlotSizeValue();
+  simulated_local_training_step_duration_in_minutes_ = operational_profiling::
+      features::GetSimulateLocalTrainingStepDurationValue();
+  ephemeral_id_lifetime_in_days_ =
+      operational_profiling::features::GetEphemeralIdLifetime();
+
+  operational_profiling_endpoint_ = GURL(federatedLearningUrl);
+
+  LoadPrefs();
+  MaybeResetEphemeralId();
+}
+
+void BraveOperationalProfiling::LoadPrefs() {
+  platform_ = GetPlatformIdentifier();
+  last_checked_slot_ = pref_service_->GetInteger(kLastCheckedSlotPrefName);
+  ephemeral_id_ = pref_service_->GetString(kEphemeralIdPrefName);
+  ephemeral_id_expiration_time_ =
+      pref_service_->GetTime(kEphemeralIdExpirationPrefName);
+}
+
+void BraveOperationalProfiling::SavePrefs() {
+  pref_service_->SetInteger(kLastCheckedSlotPrefName, last_checked_slot_);
+  pref_service_->SetString(kEphemeralIdPrefName, ephemeral_id_);
+  pref_service_->SetTime(kEphemeralIdExpirationPrefName,
+                         ephemeral_id_expiration_time_);
+}
+
+void BraveOperationalProfiling::OnCollectionSlotStartTimerFired() {
+  simulate_local_training_step_timer_->Reset();
+}
+
+void BraveOperationalProfiling::OnSimulateLocalTrainingStepTimerFired() {
+  SendCollectionSlot();
+}
+
+void BraveOperationalProfiling::SendCollectionSlot() {
+  current_collected_slot_ = GetCurrentCollectionSlot();
+  if (current_collected_slot_ == last_checked_slot_) {
+    return;
+  }
+
+  MaybeResetEphemeralId();
+
+  auto resource_request = std::make_unique<network::ResourceRequest>();
+  resource_request->url = operational_profiling_endpoint_;
+  resource_request->headers.SetHeader("X-Brave-FL-Operational-Profile", "?1");
+
+  resource_request->credentials_mode = network::mojom::CredentialsMode::kOmit;
+  resource_request->method = "POST";
+
+  url_loader_ = network::SimpleURLLoader::Create(
+      std::move(resource_request), GetNetworkTrafficAnnotationTag());
+  url_loader_->AttachStringForUpload(BuildPayload(), "application/base64");
+
+  url_loader_->DownloadToStringOfUnboundedSizeUntilCrashAndDie(
```

Comment:
perhaps `DownloadHeadersOnly` is enough?

Reply:
good point 👍